_posts/2025-03-18-movie-similarity-search-using-vector-databases.adoc

```diff
@@ -0,0 +1,437 @@
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.transaction.Transactional;
+
+@ApplicationScoped
+public class MovieRecommendationService {
+
+  @Inject
+  EmbeddingStore<TextSegment> embeddingStore;
+
+  @Inject 
+  EmbeddingModel embeddingModel;
+
+  @Transactional
+  public List<Movie> searchSimilarMovies(String overview) {
+    
+    Embedding embedding = embeddingModel.embed(overview).content();
+    EmbeddingSearchRequest request = EmbeddingSearchRequest.builder()
+    .queryEmbedding(embedding)
+    .minScore(0.5)
+    .maxResults(10)
+    .build();
+
+    return embeddingStore.search(request).matches().stream().map(m -> {
+      Long id = m.embedded().metadata().getLong("id");
+      Movie movie = Movie.findById(id);
+      return movie;
+    }).toList();
+
+  }
+}
+----
+
+
+==== Loading the movies ====
+To populate the movies in the database, we'll use a CSV file containing the top 1000 movies from IMDB.
+The important columns for us are:
+
+* `title` the movie title
+* `overview` the movie overview
+* `link` the link to the image poster
+
+To easily map CSV entries to `Movie` objects we'll use `jackson`. Specifically, we will use the `@JsonProperty` 
+annotation to map the CSV columns to the `Movie` fields. Also, we'll use `@JsonIgnoreProperties(ignoreUnknown = true)`
+to ignore unknown fields.
+
+So, a simplified version of our Movie entity object looks like: 
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+
+@Entity
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Movie extends PanacheEntity {
+
+    @JsonProperty("Poster_Link")
+    public String link;
+    @JsonProperty("Series_Title")
+    public String title;
+    @JsonProperty("Overview")
+    @Column(length = 1000)
+    public String overview;
+
+    public static List<Movie> searchByTitleLike(String title) {
+      return find("title like ?1", "%" + title + "%").list();
+    }
+}
+----
+
+So, now we are ready to load our movies from the CSV to our relational and vector databases.
+
+Key points:
+
+* Observing the `StartupEvent` allows us to load the movies when the application starts
+* CSVMapper is used to map the CSV entries to `Movie` objects
+* We save each movie using a @Transaction method because we need that `id` that's often generated by the database
+* We ingest the documents in a batch. This can make a huge difference especially, if we move from in process to remote embedding models.
```

Comment:
```suggestion
* We ingest the documents in a batch. This can make a huge difference especially if we move from in-process to remote embedding models.
```